### DISCUSSION TITLE
Show off something you've made

### DISCUSSION THREAD

**Discussion by markaren** [markaren/threepp#253] [Show and tell]
If you have created something using threepp, I'd love to see it. Please share a gif or screenshot of of your work.

**Comment by wol101** [comment]
<img width="961" alt="Screenshot 2024-04-11 132528" src="https://github.com/markaren/threepp/assets/18628692/775526ff-e73b-44fe-9d04-35dc18460ab8">
Using threepp inside a Qt OpenGL widget. It works pretty well although there are some issues with the GLRenderer when the OpenGL context gets switched, The workaround is to create a new GLRenderer each frame. I haven't got shadows to work either. Code at *[https://github.com/wol101/GaitSym5](https://github.com/wol101/GaitSym5/tree/development)*

**Reply by markaren** [reply]
Thanks for sharing. Regarding your issue, what if if you call `resetState()` on the renderer for each call? Shadows ought to work, but they often require som tuning of the shadow camera.

**Reply by wol101** [reply]
I'm sure I can get shadows working with a bit more effort but the state thing is weird. Recreating the GLRenderer works OK but calling resetState() doesn't. It effects the render and feels like maybe one of the shaders has got messed up, or some of the uniforms aren't being set to the right value (if I have a wireframe then the colour is wrong and the shape is distorted as if the MVP matrix is not correct - but I checked the material and matrices and they appear to be correct). The work around is fine for now and I really appreciate the effort that has gone into the port.

**Reply by asmwarrior** [reply]
The project is really nice! Good work. Though I'm a wxWidgets user, I still suggest if we could supply a QT example code by using this threepp library inside QT's window.

**Comment by wol101** [comment]
I’ve got a simple Qt example. It was the first thing I did. It’s based on your wxwidgets example. I’m happy to share it. Bill Sellers - Sent from my phone so please pardon my brevity.On 11 Apr 2024, at 16:05, ollydbg ***@***.***> wrote:﻿
The project is really nice! Good work. Though I'm a wxWidgets user, I still suggest if we could supply a QT example code by using this threepp library inside QT's window.

**Reply by asmwarrior** [reply]
Hi, @wol101 thanks, that's very nice. I have a code repo to share my own wxWidgets based threepp sample project, see here: https://github.com/asmwarrior/threepp/tree/wx-port, under the "cb" folder, because I use Code::Blocks as my main IDE. "threepp" is still built by cmake under msys2 mingw64.

![image](https://github.com/markaren/threepp/assets/561818/defd38fa-21f9-415f-ac64-fbab6e5cd1f9)

This is the simple demo result wxWidgets window which embed a threepp opengl scene.

**Reply by wol101** [reply]
I've created something similar for Qt at <https://github.com/wol101/threepp/tree/qt-integration> 
<img width="602" alt="Screenshot 2024-04-12 094353" src="https://github.com/markaren/threepp/assets/18628692/60d81a02-a3d5-4b4f-bc27-ccec05c6a66c">  The simple example seems to work very easily and completely as expected but once I embed it in the more complex program it only works properly the first time through the paint loop. Calling resetState leads to a mostly black screen, but recreating the renderer each time works perfectly.

**Reply by asmwarrior** [reply]
I looked at your code, and I see that you are putting the source code of the threepp directly in your project.

Also, you are using a lot of shaders(I see a lot of header files of the shaders)

I'm not familiar with the QT, so I can't help much.

**Reply by wol101** [reply]
Thanks for having a look. The performance overhead isn't bad recreating the renderer and I'll have a look at the constructor to see if I can work out what's going on. Qt uses OpenGL as its back end for drawing all its widgets so there is a fair chance that it is changing a lot of settings. I'll move threepp to a library as soon as I'm more familar with what it is doing. I find being able to easily step through the code really helpful when trying a new library.

**Comment by asmwarrior** [comment]
> Qt uses OpenGL as its back end for drawing all its widgets so there is a fair chance that it is changing a lot of settings. 

Oh, this may be the reason. Under wxWidgets, I think it use the native OS API(for example, it use win32 API under windows).

My command to build the library is here: [build-commands-under-msys2-mingw64.txt](https://github.com/asmwarrior/threepp/blob/wx-port/cb/build-commands-under-msys2-mingw64.txt), but I think I have changed the cmakefile a little to fit my need.

**Reply by wol101** [reply]
Fixed it! Qt expects the OpenGL initialisation to be done in initializeGL() but it turns out there is no valid framebuffer at that point which threepp::GLRenderer expects. I can delay creating threepp::GLRenderer till the first time the paintGL() runs and then everything behaves itself. I've updated the example code to reflect this. I've no idea why I only saw the problem in the more complex code but hopefully this now works OK. Delaying the initialisation has negligible impact on performance whereas re-initialising every paint call worried me.

**Reply by wol101** [reply]
<img width="506" alt="Screenshot 2024-04-13 181447" src="https://github.com/markaren/threepp/assets/18628692/d441156e-27d8-4db5-956e-8fa2f6283713">

Shadows too! It's almost perfect but I am getting some random crashes deep in gl::GLTextures::uploadTexture sometimes when the view is changed. This is quite a big model (>5 million vertices) but the only textures are for the shadow maps so internally generated. The errors are strange though so it could easily be some sort of memory overun in my code. I'll continue to explore.

**Comment by asmwarrior** [comment]
> <img alt="Screenshot 2024-04-13 181447" width="506" src="https://private-user-images.githubusercontent.com/18628692/322220929-d441156e-27d8-4db5-956e-8fa2f6283713.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMwNTI2MzAsIm5iZiI6MTcxMzA1MjMzMCwicGF0aCI6Ii8xODYyODY5Mi8zMjIyMjA5MjktZDQ0MTE1NmUtMjdkOC00ZGI1LTk1NmUtOGZhMmY2MjgzNzEzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDEzVDIzNTIxMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk5OWYyNjdlMjlhMmU5ODdlZTgzYWFlMWE1MTc5OTg0MGEzY2M0Y2FjZTI4NDE4M2EwOTZlNGI3OTFmMzg4NmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.DJKu41Ymo96hgCNAyc77lOCSCjBzFbUH78dkfAh00O0">
> 
> Shadows too! It's almost perfect but I am getting some random crashes deep in gl::GLTextures::uploadTexture sometimes when the view is changed. This is quite a big model (>5 million vertices) but the only textures are for the shadow maps so internally generated. The errors are strange though so it could easily be some sort of memory overun in my code. I'll continue to explore.

Nice image. You project(animal locomotion) is very interesting.

**Reply by wol101** [reply]
Thanks. I've gone back to creating the threepp::GLRenderer every paint loop. Then the shadows work perfectly. If I try and reuse the GLRenderer then gl::GLTextures::uploadTexture fails because it gets sent an invalid texture (dispose is true and it tends to crash on the dynamic_cast<DataTexture3D*> line because the pointer is not a valid texture object. I'm not sure what Qt is doing to cause this effect. It feels like it might be getting in the way of some of the dispose event housekeeping. I wonder if there something that needs to be done that is normally handled by Canvas to dispatch events...

**Reply by wol101** [reply]
Just to say that it is now working perfectly. Turns out I just wasn't disposing of things properly. Once I fixed that, I could reuse threepp::GLRenderer as expected. Someone more familiar with three.js would have spotted it straightaway. I'm very pleased with it and the performance is certainly good enough for me.

**Comment by asmwarrior** [comment]
I see the raycaster is a very nice feature. For example, if I have some points shown in the scene, when I click on one point, I would like to highlight it, and also show some text on it. Here is the screen shot I have made, you can see that I use a "circle" to highlight the selected points, also the text label is shown 

Here is the screen shots:

<img width="767" height="701" alt="image" src="https://github.com/user-attachments/assets/7630e190-8907-4de3-a34a-6d727c9c7ca3" />


See the code here:

[asmwarrior/threepp at wx-port](https://github.com/asmwarrior/threepp/tree/wx-port)
